### PR TITLE
Move artifact publishing to gating_bash_lib

### PR DESCRIPTION
**Pull request by hughsaunders**
This commit extracts two functions from the JJB-RPC-AIO_{series}-{context} job
template into gating_bash_lib. These are repo_sync and push_artifacts.

Repo sync is now a generalised rsync wrapper that could be used to push
or pull artifacts.

push_artifacts is a function specifically for pushing up artifacts
generated by an AIO job.

Connects rcbops/u-suk-dev#829

### CODE REVIEW

**Comment by hughsaunders** [comment]
Test job: http://jenkins.propter.net/job/JJB-RPC-AIO_master-swift-pr/9/

**Comment by hughsaunders** [comment]
Previous test job used the right branch, but didn't use a job that reflecte the jobs.yaml changes. Next test job: http://jenkins.propter.net/job/JJB-RPC-AIO_mitaka-13.1-cidev-ondemand/21/

**Comment by hughsaunders** [comment]
Next test build http://jenkins.propter.net/job/JJB-RPC-AIO_mitaka-13.1-cidev-ondemand/26/ (last one included a quoting fail.

**Comment by hughsaunders** [comment]
Reordered publishers so that artifact push happens before instance deletion. next test: http://jenkins.propter.net/job/JJB-RPC-AIO_mitaka-13.1-cidev-ondemand/27/

**Comment by odyssey4me** [comment]
@hughsaunders It seems that the CI job passed, but I can't see the push happening in the console log. Am I missing something?

File: scripts/gating_bash_lib.sh [hunk 1]
```diff
@@ -612,3 +612,86 @@ clone_rpc(){
...
+  # Required Environment Variables:
+  # REPO_KEY - ssh key provided by credentials binding
+  # REPO_USER - user to login to repo server as
+  # REPO_HOST - hostname/ip of repo server
+
+  # Shell Opts
+  set +x # avoid leaking ssh keys
+
+  trap "exit 0" EXIT
+  # The above trap in combination with -u will cause this script to
```
Comment (mattt416) [inline scripts/gating_bash_lib.sh]: Any harm in leaving the original `set -u` in here?  This code seems to rely on it being set so may be safer to ensure this function is setting it incase it gets removed from set_shopts_env?

Comment (mattt416) [inline scripts/gating_bash_lib.sh]: Then again, if all functions rely on it then I'd just leave it so everything is set in one place (unless we need an override, as is done with `set +x`.

**Comment by mattt416** [comment]
@odyssey4me the last test seemed to fail, so a post build wasn't triggered.  Still having a look to see what caused the failure.

**Comment by mattt416** [comment]
Ah, tempest failed, hence push_artifacts not being triggered.

**Comment by hughsaunders** [comment]
tempest.api.compute.servers.test_server_actions.ServerActionsTestJSON.test_reboot_server_hard is failing trying to login to the cirros instance with possibly the wrong password:

```
02:23:06.606     User: cirros, Password: P6@HD@b2#jKU!7Q
```

While the console output shows 
```
02:23:07.316     login as 'cirros' user. default password: 'cubswin:)'. use 'sudo' for root.
```

Not sure why its using a password at all, should be a key? Investigating.

File: scripts/gating_bash_lib.sh (hunk 1 above)
Comment (hughsaunders) [inline scripts/gating_bash_lib.sh]: I think keeping the default opts set in a central location is good, with overrides where required.

**Comment by hughsaunders** [comment]
Wasn't able to reproduce the error, so have rebased and will retest.

**Comment by hughsaunders** [comment]
previous build was lost in a jenkins cleanup, next build: http://jenkins.propter.net/job/JJB-RPC-AIO_mitaka-swift-periodic/54/

**Comment by mattt416** [comment]
@hughsaunders, it _looks_ like it did the cleanup before the pushing of artifacts?

```
...
10:55:17 + echo 'Cleanup complete'
10:55:17 Cleanup complete
10:55:17 + popd
10:55:17 /opt/jenkins_builds/workspace/JJB-RPC-AIO_mitaka-swift-periodic@3
10:55:17 [PostBuildScript] - Execution post build scripts.
10:55:17 [JJB-RPC-AIO_mitaka-swift-periodic@3] $ /usr/bin/sudo -E /bin/bash /tmp/hudson4357319906964312659.sh
10:55:17 /tmp/hudson4357319906964312659.sh: line 47: lxc-ls: command not found
10:55:17 No repo container found, quitting
```

Are `postbuildscript` items not executed in the order that they're specified?

**Comment by hughsaunders** [comment]
Last test job still ran the post build scripts out of order as the latest change hasnt taken effect. I will generate the experimental job using this PR, then run that. http://jenkins.propter.net/job/JJB-RPC-AIO_mitaka-13.1-experimental-ondemand/2/

**Comment by mattt416** [comment]
The config on http://jenkins.propter.net/job/JJB-RPC-AIO_mitaka-13.1-experimental-ondemand still shows the postbuildscripts out of order, is that expected?

**Comment by hughsaunders** [comment]
Gah, I pulled the wrong branch when I ran JJB. Have re-run with the right branch and kicked off a new job. http://jenkins.propter.net/job/JJB-RPC-AIO_mitaka-13.1-experimental-ondemand/5/  Ordering is definitely right this time.. 
screen shot 2017-01-18 at 15 10 51

File: rpc-jobs/jobs.yaml [hunk 2]
```diff
@@ -320,6 +320,13 @@
           name: IMAGE
           default: "{IMAGE}"
           description: "Pattern used to find an image to build the instance for the aio, defaults to 'Ubuntu 14.04 LTS'"
+          name: REPO_HOST
```
Comment (mattt416) [inline rpc-jobs/jobs.yaml]: Looks like we're missing the preceding "- string:" here.

**Comment by mattt416** [comment]
Failed with:

```
17:17:24 /opt/jenkins_builds/workspace/JJB-RPC-AIO_mitaka-13.1-experimental-ondemand/jenkins-rpc/scripts/gating_bash_lib.sh: line 663: REPO_USER: unbound variable
```

Do we need to set some defaults for `REPO_USER` and `REPO_HOST`?

File: rpc-jobs/jobs.yaml (hunk 2 above)
Comment (hughsaunders) [inline rpc-jobs/jobs.yaml]: Gah, yes. Aded.

**Comment by hughsaunders** [comment]
http://jenkins.propter.net/job/JJB-RPC-AIO_mitaka-13.1-experimental-ondemand/6/

**Comment by hughsaunders** [comment]
previous build failed due to missing vars, added them to ship_env_keys. Next build: http://jenkins.propter.net/job/JJB-RPC-AIO_mitaka-13.1-experimental-ondemand/7/

**Comment by hughsaunders** [comment]
Previous failed on tempest, next: http://jenkins.propter.net/job/JJB-RPC-AIO_mitaka-13.1-experimental-ondemand/9/

**Comment by hughsaunders** [comment]
Previous job passed but had reverted to the old ordering. Next attempt: http://jenkins.propter.net/job/JJB-RPC-AIO_mitaka-13.1-experimental-ondemand/16/

**Comment by hughsaunders** [comment]
Another ordering fail, re-ran jjb, checked config in the ui, looks good this time: http://jenkins.propter.net/job/JJB-RPC-AIO_mitaka-13.1-experimental-ondemand/17/

**Comment by odyssey4me** [comment]
@hughsaunders As I wasn't seeing any artifacts being published, and considering that the process needs to change anyway, I've implemented https://github.com/rcbops/rpc-artifacts/pull/28 which should publish the python artifacts in the same way for the job that executes. As such I'm not sure that the move of this function needs to happen any more. It can probably just be removed.

**Comment by hughsaunders** [comment]
Closing as not needed by @odyssey4me and further Jenkins dev will happen in this/rpc-gating

**Pull request closed without merging**